### ISSUE TITLE
Regression in 8fc5f0b * Make `jupyter-channel-ioloop` independent of `zmq`

### ISSUE THREAD

**Issue by dangom** [dzop/emacs-jupyter#141]
I noticed that after pulling to recent master I can no longer use remote REPLs.
In detail, the kernel works fine within org mode src blocks, but switching to session and trying to execute something in the REPL will fail with message "Kernel not alive" from `jupyter-repl-ret`.

After git bisecting it seems that the commit introducing the bug was

```
8fc5f0bbee9191507f594f3ab2b85adce0084c18 is the first bad commit
```

Steps to reproduce:

```
#+PROPERTY: header-args:jupyter-python :session /ssh:remote:~/kernels/remote.json

#+begin_src jupyter-python
2 + 2
#+end_src
```

Evaluate the src block, then `C-c C-v C-z` to switch to REPL and try executing commands on the REPL. You will get a message saying "Kernel not alive".

If there is any more info I could provide, I'd be happy to do so.

**Comment by dzop** [comment]
In the REPL buffer, if you evaluate `(jupyter-hb-unpause jupyter-current-client)` and then try again do you get the same error?

**Comment by dangom** [comment]
No, if I run `(jupyter-hb-unpause jupyter-current-client)` I get a long message printed on the screen and after about 2 seconds I can use the REPL without problems again.

Here's the message:

```
[nil 23834 47645 534138 nil (closure ((failed-count) (channel . #s(#s(eieio--class jupyter-hb-channel "A base class for heartbeat channels." ... [#s(cl-slot-descriptor type :hb keyword ...) #s(cl-slot-descriptor session unbound jupyter-session ...) #s(cl-slot-descriptor endpoint unbound string ...) #s(cl-slot-descriptor socket nil ... ...) #s(cl-slot-descriptor time-to-dead 1 number ...) #s(cl-slot-descriptor dead-cb ... function ...) #s(cl-slot-descriptor beating t ... ...) #s(cl-slot-descriptor paused t boolean ...)] #s(hash-table size 65 test eq rehash-size 1.5 rehash-threshold 0.8125 data (type 0 session 1 endpoint 2 socket 3 time-to-dead 4 dead-cb 5 beating 6 paused 7 ...)) nil ... [] [] #s(#5 :hb unbound unbound nil 1 ignore t t) ...) :hb #s(jupyter-session ... "592d0dde-9452-4777-adef-41b817bd4fb0" "4a084373-514bc0da642abf5e7059f668") "tcp://127.0.0.1:62410" #<user-ptr ptr=0x1368e5e80 finalizer=0x138007ed0> 1 ignore nil nil)) t) (channel-ref) (let* (... ...) (if sock ...))) (#s(hash-table size 1 test eql weakness value rehash-size 1.5 rehash-threshold 0.8125 data (t #s(#s(eieio--class jupyter-hb-channel "A base class for heartbeat channels." ... [#s(cl-slot-descriptor type :hb keyword ...) #s(cl-slot-descriptor session unbound jupyter-session ...) #s(cl-slot-descriptor endpoint unbound string ...) #s(cl-slot-descriptor socket nil ... ...) #s(cl-slot-descriptor time-to-dead 1 number ...) #s(cl-slot-descriptor dead-cb ... function ...) #s(cl-slot-descriptor beating t ... ...) #s(cl-slot-descriptor paused t boolean ...)] #s(hash-table size 65 test eq rehash-size 1.5 rehash-threshold 0.8125 data (type 0 session 1 endpoint 2 socket 3 time-to-dead 4 dead-cb 5 beating 6 paused 7 ...)) nil ... [] [] #s(#4 :hb unbound unbound nil 1 ignore t t) ...) :hb #s(jupyter-session ... "592d0dde-9452-4777-adef-41b817bd4fb0" "4a084373-514bc0da642abf5e7059f668") "tcp://127.0.0.1:62410" #<user-ptr ptr=0x1368e5e80 finalizer=0x138007ed0> 1 ignore nil nil)))) nil 0]
```

**Comment by dzop** [comment]
Hmmm...I'm not seeing how the commit you mention is related to the problem you are seeing since it doesn't have anything to do with the heartbeat channel. Your getting that "Kernel not alive" error because the heartbeat channel gets paused somewhere between the time you connect to the kernel and the time you attempt to use the REPL. This can happen if the maximum number of pings sent on the heartbeat channel has been reached before ever receiving a reply back from the kernel.

You can try increasing `jupyter-hb-max-failures` to see if you continue to get that error message. The default time between pings is 1 second and the default value of `jupyter-hb-max-failures` is 5 so if the kernel doesn't send a response to a ping within 5 seconds then this could be the cause of the problem. Maybe there was some intermittent disconnect or the connection was slow to start up?

One way to verify that you are reaching the ping limit is to put

```emacs-lisp
(jupyter-hb-on-kernel-dead
     (oref (oref jupyter-current-client kcomm) hb)
     (lambda () (message "MAX PINGS")))
```

Above the following line

https://github.com/dzop/emacs-jupyter/blob/909cdf3dd7747785acc456fc6bdc3262595126aa/jupyter-repl.el#L2223

evaluate the function and then try again to see if you get "MAX PINGS" in the message buffer when you get the "Kernel not alive" error.

**Comment by dangom** [comment]
I tried the above, but I get an error. However, after more usage I suspect you are right, this may have something to do with the quality of the connection. I'm double hopping to reach my remote (opening the ssh tunnels manually from a shell).

What suprises me is that org src blocks always work, but the REPL sometimes doesn't. I would've imagined that C-c C-c on a src block and RET on the REPL would trigger the same behaviour.

I also noticed, anecdotally, that if I wait longer before switching to a session and typing something in, that the REPL error does not appear. I'll continue testing and seeing if I can gather more information.
In any case, I cannot reproduce the issues using the commit before `8fc5f0`.

**Comment by dzop** [comment]
Sorry, I meant

```emacs-lisp
(jupyter-hb-on-kernel-dead
     (oref (oref client kcomm) hb)
     (lambda () (message "MAX PINGS")))
```
> I would've imagined that C-c C-c on a src block and RET on the REPL would trigger the same behaviour.

The hearbeat channel is not considered anywhere except in `jupyter-repl-ret` (indirectly through `jupyter-repl-connected-p`) and in `jupyter-repl-interaction-mode-line`, it probably should be considered elsewhere though.

**Comment by dzop** [comment]
>  I cannot reproduce the issues using the commit before 8fc5f0

Very odd indeed. Maybe the changes introduced in that commit revealed some other issue that has already been present.

Another thing you can try is to replace

https://github.com/dzop/emacs-jupyter/blob/909cdf3dd7747785acc456fc6bdc3262595126aa/jupyter-repl.el#L2223-L2224

with

```emacs-lisp
(prog1 (jupyter-bootstrap-repl client repl-name associate-buffer display)
  (jupyter-hb-unpause client))
```

that is, un-pause the heatbeat channel after setting up the REPL. There are a few messages passed between the kernel and Emacs when setting up the REPL which may be introducing long delays that interferes with how the heartbeat channel works.

**Comment by dangom** [comment]
Interesting news: I cannot reproduce the issue on a fast connection (working from within the network so no need to double hop today). I'll come back to this as soon as I hit it again. Very curious to understand what causes it.

**Comment by dzop** [comment]
@dangom Have you gotten a chance to see if 5b34e99 solves the problem?

**Comment by dangom** [comment]
I haven't come across the problem again. But as I said, I cannot guarantee it's solved because it wasn't always reproducible.

I'll report back if things go awry, but so far so good!! Feel free to close the issue if you think that'd be appropriate.
